tests/test_qutrit_device.py

```diff
@@ -0,0 +1,670 @@
+    probs = np.reshape(p, [3] * 3)
+    s00 = np.sum(probs[0, :, 0])
+    s10 = np.sum(probs[1, :, 0])
+    s20 = np.sum(probs[2, :, 0])
+    s01 = np.sum(probs[0, :, 1])
+    s11 = np.sum(probs[1, :, 1])
+    s21 = np.sum(probs[2, :, 1])
+    s02 = np.sum(probs[0, :, 2])
+    s12 = np.sum(probs[1, :, 2])
+    s22 = np.sum(probs[2, :, 2])
+    m_probs = np.array([s00, s10, s20, s01, s11, s21, s02, s12, s22])
+
+    marginal_test_data = [
+        (
+            np.array([0.1, 0.2, 0.3, 0.04, 0.03, 0.02, 0.01, 0.18, 0.12]),
+            np.array([0.15, 0.41, 0.44]),
+            [1],
+        ),
+        (
+            np.array([0.1, 0.2, 0.3, 0.04, 0.03, 0.02, 0.01, 0.18, 0.12]),
+            np.array([0.6, 0.09, 0.31]),
+            [0],
+        ),
+        (p, m_probs, [2, 0]),
+    ]
+
+    @pytest.mark.parametrize("probs, marginals, wires", marginal_test_data)
+    def test_correct_marginals_returned(
+        self, mock_qutrit_device_with_original_statistics, probs, marginals, wires, tol
+    ):
+        """Test that the correct marginals are returned by the marginal_prob method"""
+        num_wires = int(np.log(len(probs)) / np.log(3))  # Same as log_3(len(probs))
+        dev = mock_qutrit_device_with_original_statistics(num_wires)
+        res = dev.marginal_prob(probs, wires=wires)
+        assert np.allclose(res, marginals, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("probs, marginals, wires", marginal_test_data)
+    def test_correct_marginals_returned_wires_none(
+        self, mock_qutrit_device_with_original_statistics, probs, marginals, wires, tol
+    ):
+        """Test that passing wires=None simply returns the original probability."""
+        num_wires = int(np.log(len(probs)) / np.log(3))  # Same as log_3(len(probs))
+        dev = mock_qutrit_device_with_original_statistics(wires=num_wires)
+        dev.num_wires = num_wires
+
+        res = dev.marginal_prob(probs, wires=None)
+        assert np.allclose(res, probs, atol=tol, rtol=0)
+
+
+class TestActiveWires:
+    """Test that the active_wires static method works as required."""
+
+    def test_active_wires_from_queue(self, mock_qutrit_device):
+        queue = [
+            qml.QutritUnitary(np.eye(9), wires=[0, 2]),
+            qml.QutritUnitary(np.eye(3), wires=0),
+            qml.expval(qml.Identity(wires=5)),
+        ]
+
+        dev = mock_qutrit_device(wires=6)
+        res = dev.active_wires(queue)
+
+        assert res == Wires([0, 2, 5])
+
+
+class TestCapabilities:
+    """Test that a qutrit device defines capabilities that all devices inheriting
+    from it will automatically have."""
+
+    def test_defines_correct_capabilities(self):
+        """Test that the device defines the right capabilities"""
+        capabilities = {
+            "model": "qutrit",
+            "supports_finite_shots": True,
+            "supports_tensor_observables": True,
+            "returns_probs": True,
+            "supports_broadcasting": False,
```

Comment:
This will be supported in the future right, once decompositions and stuff are implemented? (Can note this)

Reply:
I think so. I recall one of the comments on the previous PR said that broadcasting should work out of the box. Once `DefaultQutrit` is added, I will test whether broadcasting works and update the capabilities accordingly.